### DISCUSSION TITLE
PDF generated by convert is broken after ImageMagick upgrade from 6.7.8-10 to 6.9.10-68, help required!

### DISCUSSION THREAD

**Discussion by zvereczi** [ImageMagick/ImageMagick#4368] [Help]
Hi all,

We have the following convert command working with ImageMagick 6.7.8, having 3 captions of texts and an image (logo.jpg attached as an example) on the bottom of a white background and rotated upside down by 180 degrees, see the attached test.pdf generated by 6.7.8-10. This results in a blank PDF since we upgraded to 6.9.10-68. Can someone please help to figure out what parameter changes are required for the new version of ImageMagick 6.9.10-68 to get the expected results?

convert -size 2288x555 xc:white -background white -fill black -background white -fill black -pointsize 101 -gravity Center -size 2288x290 caption:'Caption1' -size 2288x61 xc:white -background white -fill black -gravity South -pointsize 101 -size 2288x275 caption:'Caption2' -size 2288x92 xc:white logo.jpg -size 2288x31 xc:white -background white -fill black -gravity Center -pointsize 37 -size 2288x244 caption:'Caption3' -size 2288x61 xc:white -append -background white -gravity South -extent 2587x1867 -density 300x300 -colorspace CMYK -rotate 180 -gravity North -extent 2587x3734 test.pdf

The new ImageMagick/convert is:
convert -version
Version: ImageMagick 6.9.10-68 Q16 x86_64 2021-02-24 https://imagemagick.org
Copyright: © 1999-2019 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC Modules OpenMP(3.1)
Delegates (built-in): bzlib cairo fontconfig freetype gslib jng jp2 jpeg lcms ltdl lzma openexr pangocairo png ps rsvg tiff wmf x xml zlib

Any help is appreciated...

![logo](https://user-images.githubusercontent.com/92427227/137295849-4bee3ee9-d153-4331-9fe4-8e5cbd7dba89.jpg)
[test.pdf](https://github.com/ImageMagick/ImageMagick/files/7345028/test.pdf)

**Comment by fmw42** [comment]
What do you get returned from 

convert -version

Please post logo.jpg so we can use it to test your command

**Reply by fmw42** [reply]
Sorry, Ghostscript is not needed for writing to PDF

**Reply by zvereczi** [reply]
Thanks for your efforts. The blue LOGO logo.jpg is already attached, but it doesn't matter, that's not the problem. If the image is removed and we only keep the captions, the result is a same blank pdf...

**Reply by zvereczi** [reply]
convert -version
Version: ImageMagick 6.9.10-68 Q16 x86_64 2021-02-24 https://imagemagick.org
Copyright: © 1999-2019 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC Modules OpenMP(3.1)
Delegates (built-in): bzlib cairo fontconfig freetype gslib jng jp2 jpeg lcms ltdl lzma openexr pangocairo png ps rsvg tiff wmf x xml zlib

**Comment by fmw42** [comment]
I have cleaned up your code (removed extraneous duplicates, etc), organized it in logical lines of code and added +write Xx.png to view each step. I used IM 6.9.12-25. Everything is fine up to the write of PDF. I also tried changing CMYK to sRGB in case PDF writer does not like CMYK. But the result was the same -- a white page.  I note that the raster image created is upside-down and has a very long region of white at the bottom. That is most of the image is white. It looks like the PDF is not creating two pages where the first page should be the image and the second is just white. It is only showing the bottom white part.  Here is my command in Unix syntax.

```
convert \
-size 2288x555 xc:white \
\( -background white -fill black -pointsize 101 -gravity Center -size 2288x290 caption:'Caption1' +write x1.png \) \
-size 2288x61 xc:white \
\( -background white -fill black -gravity South -pointsize 101 -size 2288x275 caption:'Caption2' +write x2.png \) \
-size 2288x92 xc:white \
logo.jpg \
-size 2288x31 xc:white \
\( -background white -fill black -gravity Center -pointsize 37 -size 2288x244 caption:'Caption3' +write x3.png \) \
-size 2288x61 xc:white \
-append +write x4.png \
-background white -gravity South -extent 2587x1867 +write x5.png \
-rotate 180 +repage +write x6.png \
-gravity North -extent 2587x3734 +write x7.png \
-colorspace CMYK -density 300x300 +write x8.png \
test.pdf

```

**Reply by zvereczi** [reply]
Thanks for testing it. The white blank area is intended to be there, see the original pdf attached we'd like to achieve with the new version. Should be one page with a large empty white area at the bottom, same as in the pdf attached.

**Comment by snibgo** [comment]
After "-append", insert "+repage". But that doesn't solve the whiteness.

Before writing the output PDF, insert "-gravity None". This solves the "whiteness" (which is really transparency). I don't know why.

**Reply by fmw42** [reply]
As snibgo says, this seems to work for me by adding -gravity none or +gravity at the end

```
convert \
-size 2288x555 xc:white \
\( -background white -fill black -pointsize 101 -gravity Center -size 2288x290 caption:'Caption1' \) \
-size 2288x61 xc:white \
\( -background white -fill black -gravity South -pointsize 101 -size 2288x275 caption:'Caption2' \) \
-size 2288x92 xc:white \
logo: \
-size 2288x31 xc:white \
\( -background white -fill black -gravity Center -pointsize 37 -size 2288x244 caption:'Caption3' \) \
-size 2288x61 xc:white \
-append \
-background white -gravity South -extent 2587x1867 \
-density 300x300 -colorspace CMYK -rotate 180 +repage \
-gravity North -extent 2587x3734 \
+gravity \
test.pdf

```

**Reply by zvereczi** [reply]
Thanks, looks much better, the size and orientation seem to be correct, but I'm getting a black pdf instead of the white background if this is what you mean:
convert -size 2288x555 xc:white -background white -fill black -background white -fill black -pointsize 101 -gravity Center -size 2288x290 caption:'Caption1' -size 2288x61 xc:white -background white -fill black -gravity South -pointsize 101 -size 2288x275 caption:'Caption2' -size 2288x92 xc:white logo.jpg -size 2288x31 xc:white -background white -fill black -gravity Center -pointsize 37 -size 2288x244 caption:'Caption3' -size 2288x61 xc:white -append +repage -background white -gravity South -extent 2587x1867 -density 300x300 -colorspace CMYK -rotate 180 -gravity North -extent 2587x3734 -gravity None test.pdf

**Reply by zvereczi** [reply]
Also getting a black pdf with only the logo visible on it with the +gravity at the end, any idea why?

**Reply by fmw42** [reply]
Please format your text into readable line and new line characters and use the <> button on the panel above to preserve the exact format.  Put +write tempX.png at the end of each new image and/or at the very end before writing the pdf to see if it is a pdf issue or in your command.  Why do  you have -background white -fill black -background white. That is redundant. Put -colorspace CMYK and -density 300x300 at the end just before writing the PDF. Put +repage after -rotate.  Please try my code in my previous reply above.  If you are on Windows, we can give you the equivalent for that.

**Reply by zvereczi** [reply]
Thanks, I'm on CentOS. I'm getting a black pdf with your previous command with the +repage after rotate and +gravity at the end, only the logo is visible on it.

**Reply by fmw42** [reply]
Try this and look at the various temporary files and see where it turns black.

```
convert \
-size 2288x555 xc:white \
\( -background white -fill black -pointsize 101 -gravity Center -size 2288x290 caption:'Caption1' +write tmp1.png \) \
-size 2288x61 xc:white \
\( -background white -fill black -gravity South -pointsize 101 -size 2288x275 caption:'Caption2' +write tmp2.png \) \
-size 2288x92 xc:white \
logo: \
-size 2288x31 xc:white \
\( -background white -fill black -gravity Center -pointsize 37 -size 2288x244 caption:'Caption3' +write tmp3.png \) \
-size 2288x61 xc:white \
-append +write tmp4.png \
-background white -gravity South -extent 2587x1867 +write tmp5.png \
-density 300x300 -colorspace CMYK -rotate 180 +repage +write tmp6.png \
-gravity North -extent 2587x3734 +gravity +write tmp7.png \
test.pdf

```

**Reply by zvereczi** [reply]
Thanks, this is really useful... All seems to be fine and white until the last step. only tmp7.png is getting black...

**Reply by fmw42** [reply]
I think IM 6.9.10.68 is bad. I tried the following on my Mac and get an error.

`convert -size 100x100 xc:white white.png`

**convert: unable to open image `xc:white': No such file or directory @ error/blob.c/OpenBlob/2881.**

Do you get the same error or a black image for the result?

**Reply by fmw42** [reply]
OK. Add -background white before -gravity North in that line. Does that work?

**Reply by zvereczi** [reply]
convert -size 100x100 xc:white white.png
generates a correct 100x100 white.png image for me.

**Reply by fmw42** [reply]
OK. Add -background white before -gravity North in that line of my code for your command. Does that work?

**Reply by zvereczi** [reply]
Same result with a black tmp7.png and black pdf...

**Comment by snibgo** [comment]
As fmw42 said: Put -colorspace CMYK (and -density 300x300) at the end just before writing the PDF.

The density isn't important. The colorspace is.

**Comment by fmw42** [comment]
Try this variation. If it still is black, try changing CMYK to sRGB.

```
convert \
-size 2288x555 xc:white \
\( -background white -fill black -pointsize 101 -gravity Center -size 2288x290 caption:'Caption1' +write tmp1.png \) \
-size 2288x61 xc:white \
\( -background white -fill black -gravity South -pointsize 101 -size 2288x275 caption:'Caption2' +write tmp2.png \) \
-size 2288x92 xc:white \
logo.jpg \
-size 2288x31 xc:white \
\( -background white -fill black -gravity Center -pointsize 37 -size 2288x244 caption:'Caption3' +write tmp3.png \) \
-size 2288x61 xc:white \
-append +repage +write tmp4.png \
-background white -gravity South -extent 2587x1867 +write tmp5.png \
-rotate 180 +repage +write tmp6.png \
-background white -gravity North -extent 2587x3734 +write tmp7.png \
-colorspace CMYK -density 300x300 +write tmp8.png \
+gravity -alpha off \
test.pdf

```

**Reply by zvereczi** [reply]
MANY THANKS! This seems to work perfectly! So can you please explain why the additional +repage parameters and +gravity -alpha off at the end required for the new version comparing to our original command? Just would like to understand the changes...

**Reply by fmw42** [reply]
-extent needs to have -background white used before it to reset to white in case the default is black. Rotation always needs a +repage after it to remove any virtual canvas offsets. +gravity at the end is likely needed to keep any offsets from  getting set relative to something other than the top left.  -alpha off turns off any possible alpha channel created by your graphics/test creation that might be miss-interpreted by the PDF. Not really sure. Try removing them all and adding them in one at a time and see which one introduced the black. I suspect this version of IM had some issues/bugs that are causing this problem. It works fine on my version, so I think somewhere after your version it got fixed.

**Reply by zvereczi** [reply]
Really appreciate your time and answers. What version do you have? I also got a blank pdf with ImageMagick 7...

**Reply by fmw42** [reply]
Latest IM 6  ---  6.9.12-25

Also IM 7 --- 7.1.0-10

**Reply by zvereczi** [reply]
Thanks a lot, I'll try to update and see, but extremely happy with the current results on 6.9.10-68.

**Reply by fmw42** [reply]
Try my suggestion and see which command was the fix and report back. My suspicion is the -alpha off.
